### PR TITLE
KAFKA-14098: Add meaningful client IDs for Connect workers

### PR DESCRIPTION
**Pull request by C0urante**
[Jira](https://issues.apache.org/jira/browse/KAFKA-14098)

Adds the client ID `${groupId}-${clientId}-${topic}` to the producers and consumers used by Connect workers to interact with internal topics, where `${groupId}` is the `group.id` in the worker config file, `${clientId}` is the `client.id` in the worker config file (omitted if none is specified) and `${topic}` is one of `configs`, `statuses`, or `offsets`. Since a single shared admin client is used across all three internal topics, it is given a client ID of `${groupId}-${clientId}-shared-admin`.

### CODE REVIEW

**Comment by C0urante** [comment]
@mimaison if you have a moment, would you mind giving this a pass?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java [hunk 1]
```diff
@@ -983,7 +983,7 @@ private List<ErrorReporter> sinkTaskReporters(ConnectorTaskId id, SinkConnectorC
         if (topic != null && !topic.isEmpty()) {
             Map<String, Object> producerProps = baseProducerConfigs(id.connector(), "connector-dlq-producer-" + id, config, connConfig, connectorClass,
                                                                 connectorClientConfigOverridePolicy, kafkaClusterId);
-            Map<String, Object> adminProps = adminConfigs(id.connector(), "connector-dlq-adminclient-", config, connConfig, connectorClass, connectorClientConfigOverridePolicy, kafkaClusterId, ConnectorType.SINK);
+            Map<String, Object> adminProps = adminConfigs(id.connector(), "connector-dlq-adminclient-" + id, config, connConfig, connectorClass, connectorClientConfigOverridePolicy, kafkaClusterId, ConnectorType.SINK);
```
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Small unrelated fix spotted while I was in the neighborhood. I checked the rest of this file for similar bugs; this appears to be the only one.

File: connect/runtime/src/main/java/org/apache/kafka/connect/cli/ConnectDistributed.java [hunk 2]
```diff
@@ -106,6 +108,7 @@ public Connect startConnect(Map<String, String> workerProps) {
         // Create the admin client to be shared by all backing stores.
         Map<String, Object> adminProps = new HashMap<>(config.originals());
         ConnectUtils.addMetricsContextProperties(adminProps, config, kafkaClusterId);
+        adminProps.putIfAbsent(CLIENT_ID_CONFIG, "connect-cluster-" + config.groupId());
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/cli/ConnectDistributed.java]: With the default config, you get `connect-cluster-connect-cluster` for the client-id. I wonder if we need to add the `connect-cluster-` prefix at all?

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java [hunk 3]
```diff
@@ -1138,5 +1142,13 @@ else if (value instanceof Long)
     private String className(Object o) {
         return o != null ? o.getClass().getName() : "null";
     }
-}
 
+    private static void maybeAddClientId(Map<String, ? super String> clientProps, String groupId) {
+        String clientId = "connect-cluster";
+        if (groupId != null) {
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java]: Can `groupId` actually be null here? This should only run in distributed mode

**Review by mimaison — COMMENTED** [review]
Thanks for the PR, this is a very nice cleanup! I left a couple of questions.

File: connect/runtime/src/main/java/org/apache/kafka/connect/cli/ConnectDistributed.java (hunk 2 above)
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/cli/ConnectDistributed.java]: Yeah, that's fair. I was thinking it might be useful broker-side to call this out as a Connect application but it's not a must-have and the stuttering that you've noted with `connect-cluster-connect-cluster` is not great. Will remove.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java (hunk 3 above)
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java]: Right now we have that guarantee, but since the `*BackingStore` classes accept a `WorkerConfig` instead of a `DistributedConfig`, it's difficult to enforce that guarantee. Plus, if we ever want to use any of these backing stores with standalone mode (or a new mode that doesn't use Kafka's group membership API), it'll make that migration easier.

File: connect/runtime/src/main/java/org/apache/kafka/connect/cli/ConnectDistributed.java (hunk 2 above)
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/cli/ConnectDistributed.java]: One thing that I'm wondering a bit more about is if we should add some kind of unique identifier for each worker within the Connect cluster, since without one, it becomes harder to use broker logs to debug issues.

It's tempting to use the worker's advertised URL to identify it since that comes from the config and is likely to be human-readable. It may be a little ugly in some edge cases, though. Other possibilities are to use a hash of the advertised URL, or a UUID. Thoughts?

File: connect/runtime/src/main/java/org/apache/kafka/connect/cli/ConnectDistributed.java [hunk 4]
```diff
@@ -106,6 +108,7 @@ public Connect startConnect(Map<String, String> workerProps) {
         // Create the admin client to be shared by all backing stores.
         Map<String, Object> adminProps = new HashMap<>(config.originals());
         ConnectUtils.addMetricsContextProperties(adminProps, config, kafkaClusterId);
+        adminProps.putIfAbsent(CLIENT_ID_CONFIG, config.groupId() + "-kafka-cluster-id-lookup");
```
Comment (gharris1727) [inline connect/runtime/src/main/java/org/apache/kafka/connect/cli/ConnectDistributed.java]: nit: this adminclient seems to do a lot more than looking up the kafka cluster id. It's also responsible for creating internal topics, verifying their settings, and retrieving end-offsets. perhaps `groupId-shared-admin`? `groupId-worker-internal`?

Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/cli/ConnectDistributed.java]: Good catch; this is just a mistake. The admin client used for Kafka cluster ID lookup is instantiated elsewhere. I like `<groupId>-shared-admin`; will use that here.

File: connect/runtime/src/main/java/org/apache/kafka/connect/cli/ConnectDistributed.java (hunk 2 above)
Comment (mimaison) [reply connect/runtime/src/main/java/org/apache/kafka/connect/cli/ConnectDistributed.java]: Yeah the advertised URL is the only identifier we have for a unique worker but it could be "ugly" and pretty long. 

Ideally users would provide a unique identifier (like the pod name) as the client.id and we would prefix/suffix it with the groupId and a short name like `shared-admin`.

Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/cli/ConnectDistributed.java]: Agreed with the behavior when users explicitly supply a client ID 👍

I'd still like to give different client IDs to workers in the other case if possible. I gave base 64-encoding the advertised URL a shot and it doesn't seem too ugly.

Now when a `client.id` is supplied in the worker config, the Kafka clients for, e.g., the config backing store will have the ID `<groupId>-<userSuppliedClientId>-configs`, and when one isn't supplied, the ID will be `<groupId>-<base64EncodedAdvertisedUrl>-configs`.

How does that look?

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/ConnectUtils.java [hunk 5]
```diff
@@ -171,4 +175,28 @@ public static ConnectException maybeWrap(Throwable t, String message) {
...
+     * {@code clientIdBase(config, advertisedUrl) + "configs"} could be used as the client ID for a
+     * consumer, producer, or admin client used to interact with a worker's config topic.
+     * @param config the worker config; may not be null
+     * @param advertisedUrl the advertised URL for the worker; may not be null
+     * @return the base client ID for this worker; never null, never empty, and will always end in a
+     * hyphen ('-')
+     */
+    public static String clientIdBase(WorkerConfig config, String advertisedUrl) {
+        String clusterId = Optional.ofNullable(config.groupId())
+                .orElse("connect");
```
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/util/ConnectUtils.java:188]: We don't need coverage for this case yet since all code paths that lead to this method pass in a `DistributedConfig`, which is guaranteed to have a non-null `groupId`, but it cost little to account for it anyways so I figured we might as well.

File: connect/runtime/src/main/java/org/apache/kafka/connect/cli/ConnectDistributed.java (hunk 2 above)
Comment (mimaison) [reply connect/runtime/src/main/java/org/apache/kafka/connect/cli/ConnectDistributed.java]: Base64 can contain equal signs, I wonder if that could mess with metric names. I'm seriously considering doing nothing special if `client.id` is not there.

Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/cli/ConnectDistributed.java]: Hmm... yeah, that's fair. I'd still like to do _something_ if the user doesn't provide a `client.id`, even if it's no longer unique in a Connect cluster; could we just leave out that portion and do something like `<groupId>-configs`? Tweaked the PR to use that logic; LMKWYT

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java [hunk 6]
```diff
@@ -289,15 +292,16 @@ public static String RESTART_KEY(String connectorName) {
 
     @Deprecated
     public KafkaConfigBackingStore(Converter converter, DistributedConfig config, WorkerConfigTransformer configTransformer) {
-        this(converter, config, configTransformer, null);
+        this(converter, config, configTransformer, null, "connect-distributed");
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java]: Should this be `connect-distributed-`?

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaStatusBackingStore.java [hunk 7]
```diff
@@ -139,16 +141,17 @@ public class KafkaStatusBackingStore implements StatusBackingStore {
 
     @Deprecated
     public KafkaStatusBackingStore(Time time, Converter converter) {
-        this(time, converter, null);
+        this(time, converter, null, "connect-distributed");
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaStatusBackingStore.java]: Should this be `connect-distributed-`?

**Review by mimaison — COMMENTED** [review]
Thanks @C0urante for the updates. I think this logic makes the most sense.

**Comment by C0urante** [comment]
Thanks @mimaison, good catch with the missing hyphen suffixes in the default client IDs.

**Review by mimaison — APPROVED** [review]
LGTM, thanks @C0urante

**Comment by C0urante** [comment]
Thanks Mickael!
